wgsl/extract_grammar.pl

```diff
@@ -0,0 +1,228 @@
+sub WriteLexer($$) {
+  my ($flex_fh, $tokens_ref) = @_;
+  # Write flex file
+  foreach my $key (sort {$a cmp $b } keys %$tokens_ref) {
+    my $rule = ${$tokens_ref}{$key};
+    if (($key eq 'IDENT') or  ($key =~ m/_LITERAL/)) {
+      # Escape "
+      $rule =~ s/(")/\\$1/g;
+      $rule =~ s/MINUS/\\-/g;
+      $rule =~ s/PLUS/\\+/g;
+      $rule =~ s/PERIOD/\\./g;
+    } else {
+      # Escape characters: [ ] ( ) . * ? + { } / | ^
+      $rule =~ s/([\[\]\(\)\.\*\?\+\{\}\/\|\^])/\\$1/g;
+    }
+    print $flex_fh  "$key  $rule\n";
+  }
+  print $flex_fh "\n%%\n";
+  # The list of deferred rules.  They must go later or they may shadow
+  # a previous definition.
+  my @deferred = ();
+  foreach my $key (sort {$a cmp $b } keys %$tokens_ref) {
+    if ($key eq 'IDENT') {
+      push @deferred, "{$key}  { return $key; }\n";
+    } else {
+      print $flex_fh "{$key}  { return $key; }\n";
+    }
+  }
+  print $flex_fh @deferred;
+  # Flex has a special representation for EOF.
+  print $flex_fh "<<EOF>>  { return EOF; }\n";
+  print $flex_fh "\n%%\n";
+}
+
+sub WriteGrammar($$$) {
+  my ($bison_fh, $tokens_ref, $grammar) = @_;
+  foreach my $key (sort {$a cmp $b } keys %$tokens_ref) {
+    print $bison_fh "%token $key\n";
+  }
+  print $bison_fh "%token EOF\n";
+  print $bison_fh "\n%%\n\n";
+
+  my $rules_ref = $grammar->{'rule'};
+  foreach my $rule_key ( sort keys %$rules_ref ) {
+    print $bison_fh $rule_key, ":";
+    # get the list of right hand sides
+    my @rhs_list = @{$rules_ref->{$rule_key}};
+    my $first = 1;
+    foreach my $rhs (@rhs_list) {
+      print $bison_fh " |" unless $first;
+      $first = 0;
+      my @words = @$rhs;
+      if ($santize_re_in_rules) {
+        @words = (map { s/[()?+*]//g; $_ } (@words));
+      }
+      print $bison_fh ' ', join(' ', @words), "\n";
+    }
+    print $bison_fh "\n";
+  }
+}
+
+# Returns the list of token definitions and grammar-defining lines
+# from the WGSL spec source.
+# Input is the list of lines from the WGSL spec source.
+sub GetTokensAndGrammarLines(@) {
+  my (@input) = @_;
+  my $in_grammar = 0;
+  my $in_tokens = 0;
+  my @grammar_lines = ();
+  # Key is token name, value is regexp
+  my %token = ();
+  foreach (@input) {
+    chomp;
+    my $line = $_;
+    if ($line =~ m/<td>Token<td>/) {
+      $in_tokens = 1;
+    } elsif ($line =~ m/# Syntactic Tokens #/i) {
+      $in_tokens = 1;
+    } elsif ($in_tokens && ($line =~ m/<tr><td>`(\S+?)`<td>(.*)/)) {
+      # This is a token definition from a regular expression
+      # Preprocess into Bison/Flex form.
+      my $name = $1;
+      my $re = $2;
+      $re =~ s/\s*#.*//; # Remove trailing comment
+      $re =~ s/^`//; # Remove leading tick, if present
+      $re =~ s/`$//; # Remove trailing tick, if present
+      $re =~ s/\s+//g;  # Remove spaces
+      $token{$name} = $re;
+    } elsif ($in_tokens && ($line =~ m/<\/table>/)) {
+      $in_tokens = 0;
+    } elsif ($line =~ m/^<pre\s+class='def'/) {
```

Comment:
We can extend the class to have a 'grammar' key or something to make this more correct.